### ISSUE TITLE
Introduce arithmetic for maps?

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#1873]
Should we introduce arithmetic operators for maps?

The most common use case we have for this is:
```
residual = Map.from_geom(cmaps["counts"].geom)
residual.data = cmaps["counts"].data - npred.data
```
which could become:
```
residual = cmaps["counts"] - npred
```

The implementation could be something like:
```
def __sub__(self, other):
    # maybe some checks if geom and units are compatible here
    out = self.copy()
    out.data -= other.data
    return out
```

I think mostly we need `+` and `-` and it's OK to just take `meta` from self and not try to merge the to map meta dicts, and then it's a simple addition on the base class using something similar like the method above.

For NDData Astropy uses a mix-in class for that and it's a lot of code:
http://docs.astropy.org/en/stable/nddata/mixins/ndarithmetic.html
I'm not proposing we do something like that and spend a lot of time on this.

@registerrier @AtreyeeS - Do you think we should add it or not? If yes, would one of you be willing to implement it and send a PR?

**Comment by registerrier** [comment]
Yes why not.  But the checks are really important to avoid silent errors. 
For cases with inconsistent `MapGeom` we still have `Map.coadd`. 
Another option could be to add a test on the `MapGeom` alignement there to make it faster for identical geometries.

**Comment by cdeil** [comment]
I would not go in the direction of coadd and resampling. Let the caller be explicit for those operations.
I just meant for the common case of having two maps with the same geom.

I think basically a check like this would be sufficient and catch 99% of usage errors:
```
def _check_compatible(self, other):
    if self.data.shape != other.data.shape:
        raise ValueError("Maps data shape isn't compatible")
    if self.unit != other.unit:
        raise ValueError("Map unit isn't compatible")
```

A check on geom alignment using allclose on various things could be added as well, that's already forseen here and I think called in some tests:

https://github.com/gammapy/gammapy/blob/a486b0202b5a855e05f6d37349408a3359205468/gammapy/utils/testing.py#L128

But that isn't easy to write to cover the various geoms, so will need a bit of research and discussion and should (if at all needed) be a separate PR from the simple maps arithmetic operator PR.
